File: solutions/python/03/topological_sort/main.py
from collections import deque


def topological_sort(n: int, g: list, indeg: list) -> list:
    q = deque(i for i in range(n) if indeg[i] == 0)
    order = []
    while q:
        v = q.popleft()
        order.append(v)
        for to in g[v]:
            indeg[to] -= 1
            if indeg[to] == 0:
                q.append(to)
    return order


def main() -> None:
    n = 4
    g = [[1, 2], [3], [3], []]
    indeg = [0, 1, 1, 2]
    print(topological_sort(n, g, indeg))


if __name__ == "__main__":
    main()

File: solutions/go/03/topological_sort/main.go
package main

import "fmt"

func main() {
	n := 4
	g := [][]int{{1, 2}, {3}, {3}, {}}
	indeg := []int{0, 1, 1, 2}
	q := []int{}
	for i := 0; i < n; i++ {
		if indeg[i] == 0 {
			q = append(q, i)
		}
	}
	order := []int{}
	for head := 0; head < len(q); head++ {
		v := q[head]
		order = append(order, v)
		for _, to := range g[v] {
			indeg[to]--
			if indeg[to] == 0 {
				q = append(q, to)
			}
		}
	}
	fmt.Println(order)
}
